File: server/generate-cert.py
#!/usr/bin/env python3
"""
NeoAgroTech - Generador de certificados SSL con Python
Genera certificados SSL autofirmados usando OpenSSL
"""
import subprocess
import sys
import os
from pathlib import Path

CERT_DIR = Path(__file__).parent / 'cert'
KEY_FILE = CERT_DIR / 'server.key'
CERT_FILE = CERT_DIR / 'server.crt'

def check_openssl():
    """Verificar si OpenSSL está instalado"""
    try:
        subprocess.run(['openssl', 'version'], 
                     capture_output=True, 
                     check=True)
        return True
    except (subprocess.CalledProcessError, FileNotFoundError):
        return False

def generate_certificates():
    """Generar certificados SSL"""
    # Crear directorio si no existe
    CERT_DIR.mkdir(parents=True, exist_ok=True)
    
    # Crear .gitkeep
    gitkeep = CERT_DIR / '.gitkeep'
    if not gitkeep.exists():
        gitkeep.write_text('# Este archivo mantiene la carpeta cert en git\n')
    
    print("🔐 Generando certificados SSL autofirmados para NeoAgroTech...\n")
    
    # Verificar OpenSSL
    if not check_openssl():
        print("❌ Error: OpenSSL no está instalado o no está en el PATH.")
        print("\n💡 Instalación:")
        print("   Windows: https://slproweb.com/products/Win32OpenSSL.html")
        print("   O usa Git Bash que incluye OpenSSL")
        print("\n   Alternativa manual:")
        print(f"   1. openssl genrsa -out {KEY_FILE} 2048")
        print(f"   2. openssl req -new -x509 -key {KEY_FILE} -out {CERT_FILE} -days 365")
        return False
    
    # Verificar si ya existen
    if KEY_FILE.exists() and CERT_FILE.exists():
        print("⚠️  Los certificados ya existen.")
        print(f"   Ubicación: {CERT_DIR}")
        print("\n   Si deseas regenerarlos, elimina los archivos:")
        print(f"   - {KEY_FILE}")
        print(f"   - {CERT_FILE}")
        return True
    
    try:
        # Generar clave privada
        print("1. Generando clave privada (2048 bits)...")
        subprocess.run([
            'openssl', 'genrsa',
            '-out', str(KEY_FILE),
            '2048'
        ], check=True)
        
        # Generar certificado autofirmado
        print("2. Generando certificado autofirmado (válido por 365 días)...")
        subprocess.run([
            'openssl', 'req', '-new', '-x509',
            '-key', str(KEY_FILE),
            '-out', str(CERT_FILE),
            '-days', '365',
            '-subj', '/C=CL/ST=Chile/L=Santiago/O=NeoAgroTech/OU=IT/CN=localhost'
        ], check=True)
        
        print("\n✅ Certificados generados exitosamente!")
        print(f"   📁 Ubicación: {CERT_DIR}")
        print(f"   🔑 Clave privada: {KEY_FILE}")
        print(f"   📜 Certificado: {CERT_FILE}")
        print("\n⚠️  Estos son certificados autofirmados para desarrollo.")
        print("   Para producción, usa certificados de una CA confiable (Let's Encrypt, etc.)")
        print("\n🚀 Ahora puedes iniciar el servidor con:")
        print("   python server/https-server.py")
        print()
        return True
        
    except subprocess.CalledProcessError as e:
        print(f"\n❌ Error al generar certificados: {e}")
        return False
    except Exception as e:
        print(f"\n❌ Error inesperado: {e}")
        return False

if __name__ == '__main__':
    success = generate_certificates()
    sys.exit(0 if success else 1)

File: server/generate-cert.js
// NeoAgroTech - Generador de certificados SSL autofirmados
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const certDir = path.join(__dirname, 'cert');
const keyPath = path.join(certDir, 'server.key');
const certPath = path.join(certDir, 'server.crt');

// Crear directorio de certificados si no existe
if (!fs.existsSync(certDir)) {
  fs.mkdirSync(certDir, { recursive: true });
  // Crear archivo .gitkeep para mantener la carpeta en git
  fs.writeFileSync(path.join(certDir, '.gitkeep'), '');
}

console.log('🔐 Generando certificados SSL autofirmados para NeoAgroTech...\n');

// Verificar si OpenSSL está disponible
function checkOpenSSL() {
  try {
    execSync('openssl version', { stdio: 'ignore' });
    return true;
  } catch (error) {
    return false;
  }
}

if (!checkOpenSSL()) {
  console.error('❌ Error: OpenSSL no está instalado o no está en el PATH.');
  console.log('\n💡 Instalación:');
  console.log('   Windows: https://slproweb.com/products/Win32OpenSSL.html');
  console.log('   macOS: brew install openssl');
  console.log('   Linux: sudo apt-get install openssl (Ubuntu/Debian)');
  console.log('\n   O usa Git Bash que incluye OpenSSL.');
  process.exit(1);
}

try {
  // Verificar si los certificados ya existen
  if (fs.existsSync(keyPath) && fs.existsSync(certPath)) {
    console.log('⚠️  Los certificados ya existen.');
    console.log('   Si deseas regenerarlos, elimina los archivos en:');
    console.log(`   ${certDir}\n`);
    console.log('✅ Usando certificados existentes.');
    console.log('   Para regenerarlos, elimina los archivos y ejecuta este script nuevamente.\n');
  } else {
    generateCertificates();
  }
} catch (error) {
  console.error('\n❌ Error inesperado:', error.message);
  process.exit(1);
}

function generateCertificates() {
  try {
    // Generar clave privada
    console.log('1. Generando clave privada (2048 bits)...');
    const keyCommand = process.platform === 'win32' 
      ? `openssl genrsa -out "${keyPath}" 2048`
      : `openssl genrsa -out "${keyPath}" 2048`;
    
    execSync(keyCommand, { stdio: 'inherit' });

    // Generar certificado autofirmado
    console.log('2. Generando certificado autofirmado (válido por 365 días)...');
    const certCommand = `openssl req -new -x509 -key "${keyPath}" -out "${certPath}" -days 365 -subj "/C=CL/ST=Chile/L=Santiago/O=NeoAgroTech/OU=IT/CN=localhost"`;
    
    execSync(certCommand, { stdio: 'inherit' });

    console.log('\n✅ Certificados generados exitosamente!');
    console.log(`   📁 Ubicación: ${certDir}`);
    console.log(`   🔑 Clave privada: ${keyPath}`);
    console.log(`   📜 Certificado: ${certPath}`);
    console.log('\n⚠️  Estos son certificados autofirmados para desarrollo.');
    console.log('   Para producción, usa certificados de una CA confiable (Let\'s Encrypt, etc.)');
    console.log('\n🚀 Ahora puedes iniciar el servidor con: npm start\n');
  } catch (error) {
    console.error('\n❌ Error al generar certificados:', error.message);
    console.log('\n💡 Alternativa: Usa OpenSSL manualmente:');
    console.log(`   1. openssl genrsa -out "${keyPath}" 2048`);
    console.log(`   2. openssl req -new -x509 -key "${keyPath}" -out "${certPath}" -days 365`);
    console.log('\n   O instala OpenSSL desde:');
    console.log('   https://slproweb.com/products/Win32OpenSSL.html (Windows)');
    process.exit(1);
  }
}

